### PR TITLE
Add *.gmail.dev to trusted viewer domains

### PR DESCRIPTION
**Pull request by alabiaga**
- new domain to trusted viewers
- new opt in for untrusted xhr intercept

### CODE REVIEW

**Comment by alabiaga** [comment]
cc/ @GoTcWang @zhangsu

File: src/utils/xhr-utils.js [hunk 1]
```diff
@@ -217,7 +217,11 @@ export function getViewerInterceptResponse(win, ampdocSingle, input, init) {
       return viewer.isTrustedViewer();
     })
     .then(viewerTrusted => {
-      if (!viewerTrusted && !getMode(win).localDev) {
+      if (
+        !viewerTrusted &&
+        !getMode(win).localDev &&
+        !viewer.hasCapability('untrusted-xhr-interception')
```
Comment (choumx) [inline src/utils/xhr-utils.js]: `isExperimentOn(win, 'untrusted-xhr-interception')`

File: tools/experiments/experiments.js [hunk 2]
```diff
@@ -464,6 +464,12 @@ const EXPERIMENTS = [
     spec: 'https://github.com/ampproject/amphtml/issues/22718',
     cleanupIssue: 'https://github.com/ampproject/amphtml/issues/22749',
   },
+  {
+    id: 'untrusted-xhr-interception',
+    name: 'allow untrusted xhr interception',
```
Comment (choumx) [inline tools/experiments/experiments.js]: `'Enable "xhrInterceptor" capability for untrusted viewers. For development use only.'`

File: src/service/viewer-impl.js [hunk 3]
```diff
@@ -83,6 +83,7 @@ const TRUSTED_VIEWER_HOSTS = [
    * - www.google.cat
    */
   /(^|\.)google\.(com?|[a-z]{2}|com?\.[a-z]{2}|cat)$/,
+  /(^|\.)gmail\.dev$/,
```
Comment (choumx) [inline src/service/viewer-impl.js]: @ampproject/wg-approvers This allows Gmail's AMP4EMAIL playground to enable `xhrInterceptor` viewer capability. It provides a better (closer to prod) testing experience for email senders.

`isTrustedViewer()` is currently checked when:
- Get/set CID from viewer
- Overriding `referrerUrl` and `viewerUrl`
- Special viewer impression tracking
- Other AMP4EMAIL/AMP4ACTIONS features

Comment (cramforce) [reply src/service/viewer-impl.js]: Can we take this opportunity to get rid of this hard-coded list and instead supply it via config that is controlled by the AMP cache? Just like all the other vars in https://github.com/ampproject/amphtml/blob/master/src/config.js#L45

**Comment by alabiaga** [comment]
Sure thing, I'll make the changes. Thanks

File: src/config.js [hunk 4]
```diff
@@ -48,6 +48,19 @@ export const urls = {
...
+   * propagating the referrer. If you believe your domain should be here,
+   * file the issue on GitHub to discuss. The process will be similar
+   * (but somewhat more stringent) to the one described in the [3p/README.md](
+   * https://github.com/ampproject/amphtml/blob/master/3p/README.md)
+   *
+   * @export {!Array<!RegExp>}
+   */
+  trustedViewerHosts: [
+    /(^|\.)google\.(com?|[a-z]{2}|com?\.[a-z]{2}|cat)$/,
+    /(^|\.)gmail\.dev$/,
```
Comment (jridgewell) [inline src/config.js]: Why not just add this to the other regex?

File: src/config.js [hunk 5]
```diff
@@ -48,6 +48,19 @@ export const urls = {
...
+  /**
+   * These domains are trusted with more sensitive viewer operations such as
+   * propagating the referrer. If you believe your domain should be here,
+   * file the issue on GitHub to discuss. The process will be similar
+   * (but somewhat more stringent) to the one described in the [3p/README.md](
+   * https://github.com/ampproject/amphtml/blob/master/3p/README.md)
+   *
+   * @export {!Array<!RegExp>}
+   */
+  trustedViewerHosts: [
```
Comment (jridgewell) [inline src/config.js]: Why are we moving this to the config?

File: src/utils/xhr-utils.js (hunk 1 above)
Comment (jridgewell) [inline src/utils/xhr-utils.js]: I think this is supposed to be an OR, not an AND.

File: tools/experiments/experiments.js (hunk 2 above)
Comment (alabiaga) [reply tools/experiments/experiments.js]: done

File: src/service/viewer-impl.js (hunk 3 above)
Comment (alabiaga) [reply src/service/viewer-impl.js]: done

File: src/config.js [hunk 6]
```diff
@@ -48,6 +48,18 @@ export const urls = {
...
     env['errorReportingUrl'] || 'https://amp-error-reporting.appspot.com/r',
   localDev: env['localDev'] || false,
+  /**
+   * These domains are trusted with more sensitive viewer operations such as
+   * propagating the referrer. If you believe your domain should be here,
+   * file the issue on GitHub to discuss. The process will be similar
+   * (but somewhat more stringent) to the one described in the [3p/README.md](
+   * https://github.com/ampproject/amphtml/blob/master/3p/README.md)
+   *
+   * @export {!Array<!RegExp>}
```
Comment (choumx) [inline src/config.js]: Remove the `@export`.

File: src/config.js (hunk 4 above)
Comment (choumx) [reply src/config.js]: I also thought it was more readable as a separate regex. Please change back. 😄

**Comment by alabiaga** [comment]
Merging without @jridgewell and @cramforce review as I am away next week. @cramforce comments have been addressed, any comments from @jridgewell , I can address in a separate PR. Thanks

**Pull request merged**